### DISCUSSION TITLE
Comparison nnunetv2 versions 2.2.1 VS latest (2.7.0)

### DISCUSSION THREAD

**Discussion by hermancollin** [neuropoly/axondeepseg#979] [Ideas of new features]
In ADS, we currently have a version lock on `nnunetv2==2.2.1`. I just tried training a model with the latest version (2.7.0) and I noticed some very obvious changes in how nnunet choses hyperparameters. For example, using the exact same dataset (this one: https://doi.org/10.48324/dandi.001350/0.250511.1527), `nnUNetv2_plan_and_preprocess` choses different plans. The most surprising change is the patch size:

In 2.2.1 (current version we use in ADS):
```
"batch_size": 12,
"patch_size": [512, 512]
```
In 2.7.0:
```
"batch_size": 2,
"patch_size": [1280, 1024]
```

That's more than 4x larger! The impact on model size is modest:
- 2.2.1: 355 Mb
- 2.7.0: 453 Mb

I will compare both models and provide more details on performance (runtime, test set results).

**Comment by hermancollin** [comment]
I was able to install ADS with python 3.13.5, `nnunetv2==2.7.0` and `torch==2.7.1`. The latest version of nnunet is backwards compatible! Integrity test ran fine, even though our models use a deprecated version of nnunet plans. We do get this warning:

```
/home/ge.polymtl.ca/arcol/GRAMES/.conda/envs/ads_nnunet_latest/lib/python3.13/site-packages/nnunetv2/utilities/plans_handling/plans_handler.py:37: UserWarning: Detected old nnU-Net plans format. Attempting to reconstruct network architecture parameters. If this fails, rerun nnUNetv2_plan_experiment for your dataset. If you use a custom architecture, please downgrade nnU-Net to the version you implemented this or update your implementation + plans.
  warnings.warn("Detected old nnU-Net plans format. Attempting to reconstruct network architecture "
```

This is good news for multiple reasons:
- we no longer need to restrict python < 3.13
- we can update nnunetv2 and distribute new models without the need to retrain old models

**Comment by hermancollin** [comment]
Results of the comparison. Last 2 columns report detection metrics (only for myelinated axons). **old** means `nnunetv2==2.2.1` and **new** means `nnunetv2==2.7.0`.

| model | avg dice (axon) | avg dice (myelin) | avg F1 |
| :----------------- | :-------: | :-------: | :-------: |
| old (5-fold) | 0.971 | **0.929** | **0.949**  |
| old (light) | **0.974** | 0.928 | 0.931 |
| new (light) | 0.973 | 0.918 | 0.889 |

At a glance, the newer model seem slightly worse than previous ones, but I don't know if this holds for other datasets

**Comment by mathieuboudreau** [comment]
@hermancollin can I recommend that you formalize this  a bit more via the [seg_ci](https://github.com/axondeepseg/seg_ci) repo? Runs create outputs of all the morphometrics and masks for all ADS versions setup for the comparison, and are provided in ZIP files as run artefacts:

https://github.com/axondeepseg/seg_ci/actions/runs/24155591430
PPT comparison: https://docs.google.com/presentation/d/1JptrOQUYWxHI-2fHTRsATfLue8bP6dRcKQFe-PU8WK0/edit?usp=share_link

If so, I can help you get it setup; we'd just need a branch in ADS so there's a commit to point to the version with the newest nnunet version

**Reply by hermancollin** [reply]
@mathieuboudreau ah yes, good idea but the problem is that I'm not sure this would be very insightful, because this is not the generalist model. What I'm showing here is the stanford model that segments both myelinated and unmyelinated axons. IIRC, seg_ci tests on 2 images (one SEM and one TEM). Since the model was trained on TEM images at 0.00493 um/px, I expect it to fail on both images.

If we were to re-train the generalist model, we should definitely add to seg_ci.

**Reply by hermancollin** [reply]
btw does seg_ci report runtime?

**Reply by mathieuboudreau** [reply]
@hermancollin we could setup another v5 (nnunetv2) to run on the stanford model, we'd just need to edit how the CLI is called and the steps done beforehand (eg downloading the model). So at the very least, there would be a before and after record there.

As for runtime, I don't think there's a specific file that's saved with that, but for the versions have ADS logs I think it should be in the logfile?

**Reply by hermancollin** [reply]
@mathieuboudreau If you think we should have a record of this, we can include it in seg_ci, but wouldn't we need to add testing images more appropriate for the Stanford model?

**Reply by mathieuboudreau** [reply]
@hermancollin yup, there's a repo for that: https://github.com/axondeepseg/data_ci

**Reply by hermancollin** [reply]
@mathieuboudreau ok I'll add something there, but I think I'll need help with the seg_ci part afterward.

In the meantime, I'll add the Stanford light model to ADS so that we have a first reference point (because the one that is currently specified in model_cards.yaml is a 5-fold ensemble, so it would be inefficient to apply the model on a GitHub Actions runner.

**Comment by yolaatar** [comment]
### **!!!!! Edit: numbers in this comment were wrong, corrected below !!!!!!**


Following up on this with the resolution invariance experiment, since it gave me a good excuse to retrain the multires models with the newer nnunet and compare against the old ones.

Quick context on the models: `multires` is trained on TEM1+TEM2 combined, at multiple resolutions, fold_0. `multires_v2` is the same setup, just retrained with `nnunetv2==2.8.1` (now pinned per #996) instead of the old 2.2.1. Same data, same split, only the nnunet version changes.

## Patch size

Confirmed the same jump on our TEM1+TEM2 multires dataset:

| | patch_size | batch_size |
| :--- | :--- | :--- |
| old (2.2.1) | 512x512 | - |
| new (2.8.1) | 768x1152 | - |

Also non-square like yours (1280x1024), just a different aspect ratio, presumably because our median training image geometry differs from yours. Same story either way: nnunet's planner gives the new version a noticeably bigger field of view (roughly 2.25x more pixels per patch here).

## Aggregate Dice (TEM1 test set, fold_0, same split both versions)

| model | axon dice | myelin dice |
| :--- | :---: | :---: |
| multires (old) | 0.96-0.99 | 0.90-0.99 |
| multires_v2 (new) | 0.93-0.94 | 0.84-0.87 |

Ranges are across the resolution invariance pixel-size sweep (1.9-16 nm/px). New version is consistently a bit worse here, mirroring what you found on your own dataset (0.973/0.918 new vs 0.971-0.974/0.928-0.929 old). fold_all (`multires_full_v2`, trained on all of TEM1+TEM2) mostly closes the gap but that's not a fair comparison since it sees more data.

On the TEM2 testset (5 subjects, GT only) the two are close, and `multires_v2` actually edges out `multires_full_v2` at coarser pixel sizes.

So on pure Dice, retraining with the new nnunet doesn't look like a win, consistent with your read.

## Where it does seem to matter: patch size on a high-res image

Worth mentioning first: with the old (512x512) patch size, we'd sometimes see blank squares appear in the middle of large axons, basically a patch-shaped hole with no prediction. This happened even at lower resolutions, not just fine pixel sizes, whenever an axon was big enough relative to the patch.

To check whether the bigger patch actually helps with this, I ran both versions on a high-resolution TEM image we'd gotten some questions about recently (from William Chiesl, who reached out by email). It's in the 1 nm/px range, quite a bit finer than our training resolution and out of distribution for this model family. Same image, old vs new checkpoints, no other changes.

With the old model we get those blank squares on this high-res image. With the new version, we don't see them anymore, at least on this one.

So the bigger patch doesn't show up as a Dice improvement on our test sets, but on this real high-resolution image it does seem to help avoid a real failure mode. Might be worth checking if you see the same pattern on your end, could be a case where aggregate Dice hides a meaningful qualitative difference.

**Reply by hermancollin** [reply]
oof that's a pretty significant performance drop. I wonder if this decrease is specific to the split 0... do you think we would see such a big difference with the whole 5-fold CV? If you have these models already trained it could be interesting

**Reply by mathieuboudreau** [reply]
I'd have to see the images to understand more what's happening, @yolaatar can you share the raw + masks for both, or the different between the axon and myelin masks across both versions
?

**Reply by yolaatar** [reply]
@hermancollin Good question, we only have fold_0 and fold_all for these so can't fully rule out fold variance, but a few things point to it not just being an unlucky split:` multires_full_v2` (fold_all, no held-out split at all) shows the same drop vs the old fold_all model, the gap holds across the whole resolution sweep instead of being noisy at a couple points, and it's directionally the same as what you saw on a totally different dataset. So my guess is 5-fold CV would tighten the error bars but not flip the direction.

**Comment by yolaatar** [comment]
Update, the numbers I posted above were wrong, I messed up my metrics script, my apologies. Recomputed everything fresh, here's the corrected picture (mean Dice, TEM1 test set, native referential):

| pixel size | axon (old / new) | myelin (old / new) |
| :--- | :---: | :---: |
| 1 nm/px | 0.65 / **0.76** | 0.60 / **0.68** |
| 2.4 nm/px | 0.94 / 0.94 | 0.87 / 0.87 |
| 4.9 nm/px (train) | 0.94 / 0.94 | 0.86 / 0.86 |
| 16 nm/px | 0.94 / 0.94 | **0.86** / 0.85 |
| 30 nm/px | 0.90 / **0.91** | 0.81 / **0.82** |
| 50 nm/px | 0.46 / **0.66** | 0.44 / **0.71** |

Old and new nnUNet are basically tied across the normal resolution range (2-30nm/px). At the extremes it's actually the opposite of what I said before, new nnUNet does meaningfully better both at very fine (1nm/px) and very coarse (50nm/px) pixel sizes.

So walking back my earlier "new nnunet does worse" take. On our data, retraining with the newer version looks like a wash in the normal range and a win at the extremes, on top of fixing that patch-size artifact from before.

**Reply by yolaatar** [reply]
@hermancollin @mathieuboudreau sorry again for that :(

**Reply by hermancollin** [reply]
wow!

**Reply by mathieuboudreau** [reply]
@yolaatar no worries - this is great!
